homeassistant/components/sleepiq/__init__.py

```diff
@@ -1,115 +1,83 @@
 from homeassistant.const import CONF_PASSWORD, CONF_USERNAME, Platform
 from homeassistant.core import HomeAssistant
-from homeassistant.helpers import discovery
 import homeassistant.helpers.config_validation as cv
-from homeassistant.helpers.entity import Entity
 from homeassistant.helpers.typing import ConfigType
-from homeassistant.util import Throttle
 
-from .const import DOMAIN
-
-MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=30)
+from .const import DATA_SLEEPIQ, DOMAIN
+from .coordinator import SleepIQDataUpdateCoordinator
+from .models import SleepIQHassData
 
 _LOGGER = logging.getLogger(__name__)
 
 CONFIG_SCHEMA = vol.Schema(
     {
-        vol.Required(DOMAIN): vol.Schema(
-            {
-                vol.Required(CONF_USERNAME): cv.string,
-                vol.Required(CONF_PASSWORD): cv.string,
-            }
-        )
+        DOMAIN: {
+            vol.Required(CONF_USERNAME): cv.string,
+            vol.Required(CONF_PASSWORD): cv.string,
+        }
     },
     extra=vol.ALLOW_EXTRA,
 )
 
 
-def setup(hass: HomeAssistant, config: ConfigType) -> bool:
-    """Set up the SleepIQ component.
+PLATFORMS = [
+    Platform.BINARY_SENSOR,
+    Platform.SENSOR,
+]
 
-    Will automatically load sensor components to support
-    devices discovered on the account.
-    """
-    username = config[DOMAIN][CONF_USERNAME]
-    password = config[DOMAIN][CONF_PASSWORD]
-    client = Sleepyq(username, password)
-    try:
-        data = SleepIQData(client)
-        data.update()
-    except ValueError:
-        message = """
-            SleepIQ failed to login, double check your username and password"
-        """
-        _LOGGER.error(message)
-        return False
 
-    hass.data[DOMAIN] = data
-    discovery.load_platform(hass, Platform.SENSOR, DOMAIN, {}, config)
-    discovery.load_platform(hass, Platform.BINARY_SENSOR, DOMAIN, {}, config)
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up sleepiq component."""
+    if DOMAIN in config:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=config[DOMAIN]
+            )
+        )
 
     return True
 
 
-class SleepIQData:
-    """Get the latest data from SleepIQ."""
-
-    def __init__(self, client):
-        """Initialize the data object."""
-        self._client = client
-        self.beds = {}
-
-        self.update()
-
-    @Throttle(MIN_TIME_BETWEEN_UPDATES)
-    def update(self):
-        """Get the latest data from SleepIQ."""
-        self._client.login()
-        beds = self._client.beds_with_sleeper_status()
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the SleepIQ config entry."""
+    client = Sleepyq(entry.data[CONF_USERNAME], entry.data[CONF_PASSWORD])
+    try:
+        await hass.async_add_executor_job(client.login)
+    except ValueError:
```

Comment:
It would be better if the library raised a different exception (https://github.com/technicalpickles/sleepyq/blob/master/sleepyq/__init__.py#L182) when it can't connect so `ConfigEntryNotReady` can be raised here and the config entry can retry to setup later when the cloud service is back online.  I think @kbickar is going to replace it anyways though so it probably not worth changing at this point and it already has this problem.

Reply:
I was just looking at https://github.com/kbickar/asyncsleepiq, and it will be a different, SleepIQ specific exception. So this can be improved after switching libraries.

Reply:
I think this is also the only code missing code coverage now. And I'm starting to think that it's unnecessary to catch the ValueError here, because it should be caught in the config_flow setup instead. Does that sound right?